src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -2085,12 +2085,10 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
         validateDerivedValueClass(clazz, body3)
 
       if (!clazz.isTrait && clazz.isNonBottomSubClass(ConstantAnnotationClass)) {
-        val (pConstrOpt, auxConstrs) = body3.filter(s => s.isInstanceOf[DefDef] && s.symbol.isConstructor).splitAt(1)
-        for (p <- pConstrOpt) p.symbol.paramss match {
-          case List(ps) =>
-          case _ => ConstantAnnotationNeedsSingleArgumentList(p, clazz)
-        }
-        for (c <- auxConstrs) AuxConstrInConstantAnnotation(c, clazz)
+        val ctors = body3.iterator.collect { case method: DefDef if method.symbol.isConstructor => method }
+        val primary = ctors.next() // there is always a primary constructor
+        if (primary.symbol.paramss.lengthIs != 1) ConstantAnnotationNeedsSingleArgumentList(primary, clazz)
+        ctors.foreach(AuxConstrInConstantAnnotation(_, clazz))
```

Comment:
Nice! My only footnote (since I learned it this year) is that `isInstanceOf` doesn't checkout outer pointers, but pattern will if it can. Probably it doesn't matter here.

Reply:
TIL